flowauth/frontend/src/RoleScopePicker.jsx

```diff
@@ -2,109 +2,219 @@
+    setOpen(!open)
+  }
 
-    set_initial_state()//.catch(err => console.log(err))
-    
-  }, [role_id, server_id])
-
-  // On scopes change, jsonify the scopes into the appropriate format for
-  // MultiCascader
-  useEffect(() =>{
-    if (serverScopes.length > 0){
-      const unneeded = []
-      const baz = scopesGraph(serverScopes.map(x => x.name))
-      const foo = jsonify(baz, unneeded);
-      setRightsChoices(foo)
+  const handleCheckboxClick = () => {
+    var is_checked
+    if(isIndeterminant){
+      setIsIndeterminant(false)
+      is_checked = true
+    } else {
+      is_checked = !isChecked
     }
+    flipScopeCallback(inner_scopes.map(s => s.key), is_checked) 
+  }
+
+  return <ListItem 
+    className={classes[".MuiListItem-root"]}
+    data-cy={`nested-${outer_scope}`}
+  >
+    <div className={classes[".CollapsibleScope"]}>
+      <Checkbox 
+        checked={isChecked}
+        indeterminate={isIndeterminant}
+        onClick={handleCheckboxClick} 
+        data-cy={"checkbox"}
+      />
+      <ListItemText primary = {outer_scope} className={classes[".MuiTextItem-root"]} />
+      <Button onClick={handleChevronClick} data-cy={"chevron"}>
+        {open ? <ExpandLess className={classes[".MuiSvgIcon-root"]} /> : <ExpandMore className={classes[".MuiSvgIcon-root"]} />}
+      </Button>
+    </div>
+    <Collapse in={open}>
+      <ScopeList scopes = {inner_scopes} flipScopeCallback={flipScopeCallback}/>
+    </Collapse> 
+  </ListItem>
+}
 
-  }, [serverScopes])
+
+function ScopeList (props) {
+  const classes = useStyles();
+  const {scopes, flipScopeCallback} = props
+  const [flatScopes, setFlatScopes] = useState([])
+  const [nestedScopes, setNestedScopes] = useState([])
 
   useEffect(() => {
-    updateScopes(roleScopes)
-  }, [roleScopes])
-  
-  const handleChange = (checked_roles) => {
-    //'checked_roles' in this case is a list of _labels_ from RightsCascade's onChange
-    // So we need to filter RoleScopes to all the scopes with that 'name' string
-    // This is further complicated by RightsCascade...apparently only returning the root
-    // label of a tree? This could be a case of it returning the lowest common root of a
-    // selected node; so if there is _only_ admin3:dummy_query:dummy_query, and you select
-    // admin3 -> dummy_query -> dummy_query, you have selected all _admin3_, therefore 
-    // new_labels contains only admin3.
+    const flat_scopes = scopes.filter(s => !s.name.includes(":"))
+    setFlatScopes(flat_scopes)
     
-    setSelectedRights(checked_roles)
-    console.debug("checked_roles:",checked_roles)
-    setRoleScopes(serverScopes.filter(
-      s_scope => checked_roles.some(
-        c_role => s_scope.name.startsWith(c_role)
-      )
-    ))
-  };
-
-  return (
-    <RightsCascade
-      options={rightsChoices}
-      value={selectedRights}
-      onChange={handleChange}
-      // label={"Scopes"}
-    />
-  );
+    // This mess takes every scope that is 
```

Comment:
Easier to just return that form from the backend?

Reply:
It needs both fully-resolved scope paths and the grouped rep; it'd need to be mangling whatever's returned in one direction or the other.